airflow/models/taskinstance.py

```diff
@@ -539,7 +539,7 @@ def _refresh_from_db(
         task_instance.end_date = ti.end_date
         task_instance.duration = ti.duration
         task_instance.state = ti.state
-        task_instance.try_number = ti._try_number  # private attr to get value unaltered by accessor
+        task_instance.try_number = _get_private_try_number(task_instance=ti)
```

Comment:
I wonder if this can be improved when we implement TaskInstanceRun…

Reply:
Yes this problem should go away completely 

I actually doubt we really need these shenanigans now (specifically the conditional +1 logic)

Reply:
I don’t know either (never fully understood the logic tbh)

Reply:
According to the comments it is just “to ensure logs go to right file “ but if that’s really all, it could be corrected in the file task handler and try number could just be one value

Reply:
I think there were at least one or two attempts I remember in the past to remove it - and failed so far. 

I believe there is race conditions that try number is increased in the DB when task state is changed from RUNNING to FAILED (for example when user stops the task from the UI)  and try_number is already stored in the DB - but the running task might still be running and stopping itself, so it should continue logging to the same try_number it was started with.